src/token/fa2.mligo

```diff
@@ -10,8 +10,16 @@
 // -----------------------------------------------------------------
 
 [@inline]
-let check_sender (from_ , store : address * storage): address =
-  if (sender = store.admin) then from_
+let validate_token_type (token_id, store : token_id * storage): token_id =
+  if Big_map.mem token_id store.total_supply
+  then token_id
+  else
+    ([%Michelson ({| { FAILWITH } |} : string * unit -> token_id)]
+     ("FA2_TOKEN_UNDEFINED", ()) : token_id)
+
+[@inline]
+let check_sender (from_, token_id, store : address * nat * storage): address =
+  if (sender = store.admin && token_id = store.frozen_token_id) then from_
   else if (sender = from_) then from_
```

Comment:
Ah this check is kinda sad because (while it makes sense) it is repeated in `transfer_one`:
```ocaml
        if tx.token_id = store.frozen_token_id then (
          if sender = store.admin then tx.token_id
```
I think we can revert this and only call `check_sender` when we need to, aka when the `token_id` is not `frozen_token_id`.

So, my suggestion is to here just do:
```suggestion
let check_sender (from_, store : address * storage): address =
  if (sender = from_) then from_
```
and change the implementation of `transfer_item` to:
```ocaml
let transfer_item (store, ti : storage * transfer_item): storage =
  let transfer_one (store, tx : storage * transfer_destination): storage =
    if tx.amount = 0n then store
    else
      let (valid_token_id, valid_from_) =
        if tx.token_id = store.frozen_token_id
        then
          if sender = store.admin
          then (tx.token_id, ti.from_)
          else (failwith("FROZEN_TOKEN_NOT_TRANSFERABLE") : token_id * address)
        else (validate_token_type (tx.token_id, store), check_sender (ti.from_, store))
      in
      let ledger = store.ledger in
      let total_supply = store.total_supply in
      let (ledger, total_supply) = debit_from(tx.amount, valid_from_, valid_token_id, ledger, total_supply) in
      let (ledger, total_supply) = credit_to(tx.amount, tx.to_, valid_token_id, ledger, total_supply) in
      { store with
        ledger = ledger
      ; total_supply = total_supply
      }
  in List.fold transfer_one ti.txs store
```

Reply:
I like the tuple trick, much cleaner :+1:

Now that I think about it once again, I see some problems still remain:

1. FA2 states that:
> If one of the specified `token_id`s is not defined within the FA2 contract, the entrypoint MUST fail with the error mnemonic `"FA2_TOKEN_UNDEFINED"`.
However, we check for zero transfer before validating the token id (in my current implementation as well). One may attempt to transfer zero to check if a certain token id is present, and we always succeed in this case.

2. The "`token_id` in `total_supply`" check is done three times: in `validate_token_type`, in `debit_from`, and in `credit_to`.
    - I can't remove the check from `debit_from` because `debit_from` actually updates the total supply (and we need to fail if it's `None`).
    - If I remove `validate_token_type`, I'll get `FA2_INSUFFICIENT_BALANCE` instead of `FA2_TOKEN_UNDEFINED`. I can adjust `debit_from` so that it first tries checks the total supply – I think this would be the best option.
    - I can, and I think I **should** remove this check from `credit_to`: it's extraneous in `transfer`, and it's plain wrong in `mint`: some DAO authors may want to mint new tokens with custom token ids, and mint uses `credit_to`.

So I'd rather:
- remove the `tx.amount = 0n` checks
- remove `validate_token_type`
- adjust the code in `debit_from` so that it checks for token type
- treat nonexisting total supply in `credit_to` as zero supply

Reply:
> we check for zero transfer before validating the token id

This has often been a point of contention, in FA1/1.2 as well, although I think it's pretty clear in this case, given also the:
> Transfers of zero amount MUST be treated as normal transfers.

So regarding point 1, I also think we should remove the `tx.amount = 0n` check entirely.

Regarding point 2 instead:
> remove `validate_token_type`

Well we need that in other places as well, but if you only mean in `transfer_one` then yes, it makes sense to me as well :ok_hand: 

> adjust the code in `debit_from` so that it checks for token type

Likewise this also seem a good move, as you said it should be possible to get by moving the check up.

> treat nonexisting total supply in `credit_to` as zero supply

This I think is not correct tho: just as `debit_from` does, the `credit_to` function needs to update the total supply and to do so it requires the `current_total_supply`.
This means that the "check" doesn't really add a cost, as the `Map.find_opt` is needed anyway and I also think that it should fail when trying to increase the supply of an undefined token.

Note: I also think that should be the case for `mint` and not allow admins to add `token_id`s, moreover it seems that we'll likely remove the `mint` entirely, so let's keep this behavior.